crates/ntx-builder/src/builder/data_store/mod.rs

```diff
@@ -0,0 +1,162 @@
+use std::{collections::BTreeSet, sync::Arc};
+
+use account_cache::NetworkAccountCache;
+use miden_node_proto::domain::account::{NetworkAccountError, NetworkAccountPrefix};
+use miden_objects::{
+    AccountError, MastForest, Word,
+    account::{Account, AccountId},
+    block::{BlockHeader, BlockNumber},
+    crypto::merkle::PartialMmr,
+    note::{NoteScript, NoteTag},
+    transaction::{ExecutedTransaction, PartialBlockchain},
+};
+use miden_tx::{DataStore, DataStoreError, MastForestStore, TransactionMastStore};
+use thiserror::Error;
+use tokio::sync::Mutex;
+use tracing::warn;
+
+use super::store::{StoreClient, StoreError};
+use crate::COMPONENT;
+
+mod account_cache;
+
+// DATA STORE
+// =================================================================================================
+
+pub struct NtxBuilderDataStore {
+    mast_forest_store: TransactionMastStore,
+    store_client: StoreClient,
+    account_cache: NetworkAccountCache,
+    block_ref: Mutex<BlockHeader>,
+    partial_mmr: Mutex<PartialMmr>,
+}
+
+impl NtxBuilderDataStore {
+    pub async fn new(store_client: StoreClient) -> Result<Self, NtxBuilderError> {
+        let account_cache = NetworkAccountCache::new(128);
+        let mast_forest_store = TransactionMastStore::new();
+
+        // SAFETY: ok to unwrap because passing `None` should return the latest data everytime
+        let (block_ref, partial_mmr) =
+            store_client.get_current_blockchain_data(None).await?.unwrap();
+
+        Ok(Self {
+            account_cache,
+            mast_forest_store,
+            store_client,
+            block_ref: Mutex::new(block_ref),
+            partial_mmr: Mutex::new(partial_mmr),
+        })
+    }
+
+    pub fn insert_note_script_mast(&self, note_script: &NoteScript) {
+        self.mast_forest_store.insert(note_script.mast());
+    }
+
+    pub async fn get_cached_acc_or_fetch_by_tag(
+        &self,
+        note_tag: NoteTag,
+    ) -> Result<Option<Account>, NtxBuilderError> {
+        let account_prefix: NetworkAccountPrefix = note_tag.try_into().unwrap();
+        // look in cache, try the store otherwise
+        let account = if let Some(acc) = self.account_cache.get(account_prefix) {
+            Some(acc)
+        } else if let Some(acc) = self.store_client.get_network_account_by_tag(note_tag).await? {
+            // insert to cache
+            self.account_cache.put(&acc)?;
+            Some(acc)
+        } else {
+            None
+        };
+
+        if let Some(acc) = &account {
+            self.mast_forest_store.insert(acc.code().mast());
+        }
+
+        Ok(account)
+    }
+
+    pub async fn update_blockchain_data(&self) -> Result<BlockNumber, StoreError> {
```

Comment:
As mentioned in one of the other comments, in the future, we could probably just add block headers to the state once they are committed to the store. So, this could become something like:
```rust
pub fn add_block_header(&mut self, block_header: BlockHeader) -> Result<BlockNumber, SomeError> {
    ...
}
```
Internally, it would update the `partial_mmr` and `block_ref` fields (similar to how does now).